clang/lib/CIR/CodeGen/CIRGenExpr.cpp

```diff
@@ -543,12 +543,6 @@ void CIRGenFunction::buildStoreOfScalar(mlir::Value Value, Address Addr,
                                         bool Volatile, QualType Ty,
                                         LValueBaseInfo BaseInfo, bool isInit,
                                         bool isNontemporal) {
-  if (!CGM.getCodeGenOpts().PreserveVec3Type) {
```

Comment:
This looks a bit tricky to me. By removing this, how do we still track `CGM.getCodeGenOpts().PreserveVec3Type` behaviors from when it's ON? I'd expect an assert for the case you are currently not handling, but this seems to silently continue?

Reply:
From browsing through the Clang source code, it looks like Clang will optimize loads and stores of vectors of length three to be loads and stores of length four instead.  The command-line option `-Xclang -fpreserve-vec3-type` will disable that optimization.

The ClangIR path doesn't do that optimization.  So `CGM.getCodeGenOpts().PreserveVec3Type` should have no effect, since there is no optimization to skip.

I don't know how to test this right now, because GNU vector types don't allow a vector of length three, and GNU vectors are the most accessible on my Linux x86 machine.  It must be one of the several other kinds of vector types that gives special meaning to a 3-length vector.

But you point about how to track this is good.  I'll add an assert that the length is not three or PreserveVec3Type is true, because a length-three vector without that option is when we might need to do something different.

Reply:
> But you point about how to track this is good. I'll add an assert that the length is not three or PreserveVec3Type is true, because a length-three vector without that option is when we might need to do something different.

Sounds great. Since we had no vector support, we just assert'ed at the coarse level, but one improvement here (that could allow us to track these specific behaviors) is to abide a bit more to the traditional codegen skeleton (as in both `EmitLoadOfScalar` and `EmitStoreOfScalar`), for example:

```
buildLoadOfScalar(... {
...
if (const auto *ClangVecTy = Ty->getAs<VectorType>()) {
    // Boolean vectors use `iN` as storage type.
    if (ClangVecTy->isExtVectorBoolType()) {
      llvm_unreacheable("NYI");
    }

    // Handle vectors of size 3 like size 4 for better performance.
    const llvm::Type *EltTy = Addr.getElementType();
    const auto *VTy = cast<llvm::FixedVectorType>(EltTy);

    if (!CGM.getCodeGenOpts().PreserveVec3Type && VTy->getNumElements() == 3) {
      llvm_unreacheable("NYI");
    }
  }
```

By just following the more complete skeleton I believe you might get the same behavior as deleting the code (while preserving the entry points for future incremental work)

Reply:
I have added back in the checks for `PreserveVec3Type`, but with more conditions so the failures are more precise.